dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,351 @@
+{
+    StepValue(true);
+}
+void NumberBox::OnNumberBoxKeyUp(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const& args)
+{
+    switch (args.Key()) {
+        case winrt::VirtualKey::Up:
+        case winrt::VirtualKey::GamepadDPadUp:
+            StepValue(true);
+            break;
+        case winrt::VirtualKey::Down:
+        case winrt::VirtualKey::GamepadDPadDown:
+            StepValue(false);
+            break;
+    }
+}
+void NumberBox::OnScroll(winrt::IInspectable const& sender, winrt::PointerRoutedEventArgs const& args) {
+    if (!HyperScrollEnabled()) {
+        return;
+    }
+    int delta = args.GetCurrentPoint(*this).Properties().MouseWheelDelta();
+
+    if (delta > 0)
+    {
+        StepValue(true);
+    }
+    else if (delta < 0)
+    {
+        StepValue(false);
+    }
+
+}
+
+
+// Increments or decrements value by StepFrequency, wrapping if necessary
+void NumberBox::StepValue(bool sign)
+{
+    // Validating input before and after
+    ValidateInput();
+    double newVal = Value();
+
+    if (sign)
+    {
+        newVal = newVal + StepFrequency();
+    }
+    else
+    {
+        newVal =  newVal - StepFrequency();
+    }
+
+    // MinMaxMode Wrapping
+    if (MinMaxMode() == winrt::NumberBoxMinMaxMode::WrapEnabled && !IsInBounds(newVal))
+    {
+        while ( newVal > MaxValue() )
+        {
+            newVal = MinValue() + (newVal - MaxValue()) - 1;
+        }
+        while ( newVal < MinValue() )
+        {
+            newVal = MaxValue() - abs(newVal - MinValue()) + 1;
+        }
+        Value(newVal);
+        UpdateTextToValue();
+        return;
+    }
+
+    // Update Text and Revalidate new value
+    Value(newVal);
+    UpdateTextToValue();
+    ValidateInput();
+
+}
+
+
+
+
+// Runs formatter and updates TextBox to it's value property, run on construction if Value != 0
+void NumberBox::UpdateTextToValue()
+{
+        winrt::hstring formattedValue(m_formatter.Format(Value()));
+        Value( (m_formatter.ParseDouble(formattedValue)).Value() );
+        m_TextBox.Text(formattedValue);
+}
+
+// Handlder for swapping visual states of textbox
+// TODO: Implement final visual states in spec
+void NumberBox::SetErrorState(bool state)
+{
+    if (state)
+    {
+        m_hasError = true;
+        winrt::VisualStateManager::GoToState(*this, L"Invalid", false);
+    }
+    else
+    {
+        m_hasError = false;
+        winrt::VisualStateManager::GoToState(*this, L"Valid", false);
+    }
+
+}
+
+// Enables or Disables Spin Buttons
+// TODO: Styling for buttons
+void NumberBox::SetSpinButtonVisualState()
+{
+
+    if ( SpinButtonPlacementMode() == winrt::NumberBoxSpinButtonPlacementMode::Inline )
+    {
+        winrt::VisualStateManager::GoToState(*this, L"SpinButtonsVisible", false);
+    }
+    else if ( SpinButtonPlacementMode() == winrt::NumberBoxSpinButtonPlacementMode::Hidden )
+    {
+        winrt::VisualStateManager::GoToState(*this, L"SpinButtonsCollapsed", false);
+    }
```

Comment:
do you need to update all the properties of the formatter when just one of the control's properties change ? If IntegerDigits change do you need to update FrationDigits etc ? Would it make sense to have individual callbacks and just update the properties that are changing?